providers/MetaMaskWalletProvider.tsx

```diff
@@ -0,0 +1,102 @@
+import { useMetaMaskExtension } from "@/providers/MetaMaskExtenstionProvider";
+import {
+	createContext,
+	FC,
+	useCallback,
+	useContext,
+	useEffect,
+	useState,
+} from "react";
+import { ethers } from "ethers";
+import { MetaMaskAccount } from "@/types";
+import { ensureEthereumChain } from "@/utils";
+
+interface MetaMaskWalletContextType {
+	connectWallet: (callback?: () => void) => Promise<void>;
+	selectedAccount: MetaMaskAccount;
+	wallet: ethers.providers.Web3Provider;
+}
+
+const MetaMaskWalletContext = createContext<MetaMaskWalletContextType>(
+	{} as MetaMaskWalletContextType
+);
+
+interface MetaMaskWalletProviderProps {}
+
+const MetaMaskWalletProvider: FC<MetaMaskWalletProviderProps> = ({
+	children,
+}) => {
+	const { extension, promptInstallExtension } = useMetaMaskExtension();
+	const [wallet, setWallet] =
+		useState<MetaMaskWalletContextType["wallet"]>(null);
+	const [selectedAccount, setSelectedAccount] =
+		useState<MetaMaskWalletContextType["selectedAccount"]>(null);
+
+	const connectWallet = useCallback(
+		async (callback) => {
+			if (!extension) {
+				callback?.();
+				return promptInstallExtension();
+			}
+
+			await ensureEthereumChain(extension);
+
+			const accounts = (await extension.request({
+				method: "eth_requestAccounts",
+			})) as string[];
+
+			if (!accounts?.length)
+				return alert(
+					"Please create at least one account in MetaMask extension to continue."
+				);
+
+			setSelectedAccount({ address: accounts[0] });
+			setWallet(new ethers.providers.Web3Provider(extension as any));
+		},
+		[extension, promptInstallExtension]
+	);
+
+	useEffect(() => {
+		if (!extension) return;
+		const checkAccounts = async () => {
+			const accounts = (await extension.request({
+				method: "eth_accounts",
+			})) as string[];
+			if (!accounts?.length) return;
+
+			setSelectedAccount({ address: accounts[0] });
+			setWallet(new ethers.providers.Web3Provider(extension as any));
+		};
+
+		checkAccounts();
+	}, [extension]);
+
+	useEffect(() => {
+		if (!selectedAccount?.address || !extension) return;
+
+		const onAccountsChanged = (accounts: string[]) => {
+			if (!accounts?.length) return setSelectedAccount(null);
+			setSelectedAccount({ address: accounts[0] });
+		};
+
+		extension.on("accountsChanged", onAccountsChanged);
+
```

Comment:
i like writing the callback inside the call to `on`, is this your preference?

good time to bring up the question of what's better, defining async function in useEffect then calling, or `(async () => {})()`

Reply:
You mean an inline callback like this? `extension.on("accountsChanged", () => {});`, yeah I don't mind that either, the only reason I goes with the above is because I have to have a reference of that function so I can do `extension.removeEventListener` later

All in all tho, it's still better to have a proper named function, because we can (potentially) pick up the name out it when there is an error. So it's best to always do `const onAccountsChanged = () => {}`, and it's also best to avoid anonymous function call like this `(async () => {})()`. I dig get lazy in some cases tho, so gonna update that too.